### DISCUSSION TITLE
Wallet connect v2 causing an infinite loop when switching networks (Trust wallet)

### DISCUSSION THREAD

**Discussion by mcoetzee** [WalletConnect/walletconnect-monorepo#2673] [Developer Support]
### What is happening?
When connecting on desktop with a Trust mobile wallet (by scanning the QR code), the v2 connector gets stuck in an infinite loop after switching networks. The _really weird_ part is that it only gets stuck in this infinite loop if I render a `Tooltip` component (`@radix-ui/react-tooltip`).

### Steps to reproduce
**Reproduction repo**: https://github.com/mcoetzee/wallet-connect-bug/tree/fa858c2d1bcc06bccb2e0a759c90f968486f4235/
After running:
```bash
yarn
yarn start
```

1. Connect with mobile Trust wallet by scanning QR code (Avalanche will be connected network)
2. Switch network to Ethereum (this will work as expected)
3. Check the checkbox with label "Check this and switch network to see infinite loop"
4. Switch network to Avalanche (or Polygon)
5. You should then see in the console the connected chain endlessly switching back and forth:
```
>> new chain detected: 43114
>> new chain detected: 1
>> new chain detected: 43114
>> new chain detected: 1
>> new chain detected: 43114
>> new chain detected: 1
....
```
You can see me reproducing the issue in this recording as well:

https://github.com/WalletConnect/walletconnect-monorepo/assets/747979/17b74851-b70f-4e47-aa43-39dcded99a0a

### Dependencies
We're using `wagmi` v0.12.16 because migrating from `ethers` to `viem` isn't an option at the moment.
```
"wagmi": "0.12.16"
"@web3modal/ethereum": "2.4.7"
"@web3modal/react": "2.4.7"

"react": "18.2.0"
"react-dom": "18.2.0"
"react-scripts": "5.0.1"

"@radix-ui/react-popover": "1.0.6"
"@radix-ui/react-tooltip": "1.0.6"
```

### Related bug
I have also noticed that when connecting on desktop with a Trust mobile wallet (by scanning the QR code), the v2 connector switches back and forth between two networks a number times in a row. The _really weird_ part is that it only does this if I render a `Popover` component (`@radix-ui/react-popover`) after connecting.

### Steps to reproduce
**Using the same reproduction repo**: https://github.com/mcoetzee/wallet-connect-bug/tree/fa858c2d1bcc06bccb2e0a759c90f968486f4235/

1. Connect with mobile Trust wallet by scanning QR code (Avalanche will be connected network)
2. Switch network to Ethereum (this will work as expected)
3. Refresh the browser
4. Click the Connect button (this should automatically connect you to the Trust wallet)
5. You should then see in the console the connected chain switching back and forth a number of times:
```
>> new chain detected: 43114
>> new chain detected: 1
>> new chain detected: 43114
>> new chain detected: 1
>> new chain detected: 43114
>> new chain detected: 1
....
```
You can see me reproducing the issue in this recording as well:
    
https://github.com/WalletConnect/walletconnect-monorepo/assets/747979/0e97a91e-553f-40fd-84b1-2153c5809d31

**Comment by xzilja** [comment]
There are a lot of angles to look at here, so some ideas:

1. Only wagmi v1 is compatible with latest versions of the modal, if you need to stay on `0.12.x` of wagmi, you'll likely need to downgrade to modal's version `2.2.x`.

2. I'd figure out what is causing re-rendering, likely that `Popover` component you mentioned, as this is not web3modal related, nor do I think wagmi hook is reporting wrong data (as it is being used widely without this issue).

3. Test against https://lab.web3modal.com (it's [open source](https://github.com/WalletConnect/web3modal)) to see if you can replicate and find potential leads.

**Reply by mcoetzee** [reply]
I did a lot of digging and could see that the [watchSigner function's](https://github.com/wagmi-dev/wagmi/blob/0.12.x/packages/core/src/actions/accounts/watchSigner.ts#L19) `handleChange` listener was firing over and over again (as if the listener doesn't get unsubscribed after switching chains`).
<img width="745" alt="Screenshot 2023-06-19 at 12 12 29" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/747979/b9cee6e3-eb38-4010-b13b-f761f943801d">

FYI downgrading did not fix it. Another option I have is to upgrade to `wagmi` v1 and use their Ethers.js adapters: https://wagmi.sh/react/ethers-adapters

**Reply by xzilja** [reply]
I've never seen this happen before (old and new wagmi versions) so not sure what to suggest. I'd try common tricks like checking `useEffects` that can be looping (ones that might have objects or other deeply nested values passed to dependency array). I also noticed that you import from `wagmi` and `@wagmi/core` in your codebase, while it shouldn't be causing this, better use `wagmi` directly. If you installed `@wagmi/core`, remove that one and rely on one that comes with that specific version of wagmi.

**Reply by mcoetzee** [reply]
Yeah this is probably top 10 weirdest issues I've come across in my whole career. Thinking I will probably upgrade and take the Ethers.js adapter approach to solve this.

FYI I've repurposed that repo for another issue I have (https://github.com/orgs/WalletConnect/discussions/2694), so if you wanted to look at the code that reproduces this weird infinite loop bug, then you can look at the code at this commit: https://github.com/mcoetzee/wallet-connect-bug/tree/fa858c2d1bcc06bccb2e0a759c90f968486f4235/

> I also noticed that you import from wagmi and @wagmi/core in your codebase

Yip that is the code for the Ethers.js adapters, that I got from here: https://wagmi.sh/react/ethers-adapters
